src/odemis/driver/technolution.py

```diff
@@ -0,0 +1,544 @@
+                                               setter=self._setScanDelay)
+
+        self._metadata[model.MD_PIXEL_SIZE] = self.pixelSize.value
+        self._metadata[model.MD_DWELL_TIME] = self.dwellTime.value
+
+    def _setResolution(self, input_resolution):
+        if input_resolution[0] == input_resolution[1]:
+            return input_resolution
+        else:
+            logging.warning("Non-square resolution entered, only square resolution sizes are supported")
+            return self.resolution.value
+
+    def _setpixelSize(self, input_resolution):
+        if input_resolution[0] == input_resolution[1]:
+            return input_resolution
+        else:
+            logging.warning("Non-square pixel size entered, only square pixel sizes are supported")
+            return self.pixelSize.value
+
+    def _setScanDelay(self, input_delay):
+        if not(hasattr(self.parent, "_MPPC")) or self.parent._MPPC.acqDelay.value - input_delay[0] >= 0:
+            return input_delay
+        else:
+            logging.error("Wrong input x_scan_to_acq_delay got a negative value.\n"
+                          "Change values so that 'self.parent._MPPC. acqDelay.value - self.scanDelay.value[0]' has a positive result")
+            return self.scanDelay.value
+
+class MirrorDescanner(model.Emitter):
+
+    def __init__(self, name, role, parent, **kwargs):
+        super(MirrorDescanner, self).__init__(name, role, **kwargs)
+        self.parent = parent
+        # model.Emitter(name, role)
+        self.rotation = model.FloatContinuous(0, range=(0, 2 * numpy.pi), unit='rad, cw=+')
+        self.scanOffset = model.TupleContinuous((0.0, 0.0), range=((-10.0, -10.0), (10.0, 10.0)), unit='V,V')
+        self.scanGain = model.TupleContinuous((0.0, 0.0), range=((-10.0, -10.0), (10.0, 10.0)), unit='V,V')
+
+
+class MPPC(model.Detector):
+    def __init__(self, name, role, parent, **kwargs):
+        super(MPPC, self).__init__(name, role, **kwargs)
+        self.parent = parent
+
+        if not (hasattr(self.parent, "_MirrorDescanner")) or not (hasattr(self.parent, "_EBeamScanner")):
+            ConnectionError("Not all required children are defined (maybe definition order is not correct)")
+
+        self._field_data = FieldMetaData(None, None)  # Current field_data
+
+        # VA defauls values
+        self._shape = model.TupleVA((8, 8, 65536), unit="cell-images and depth")
+        self.path = model.StringVA('~/Pictures')
+        self.dataContent = model.StringVA('empty')
+        self.filename = model.StringVA(time.strftime("default-%H-%M-%S-%Y-%m-%d"))
+        self._server_URL = self.parent.server_URL
+        self.externalStorageURL_host = model.StringVA('host')
+        #NOTE: Do not write real username/password here since this is published on github in plaintext!
+        self.externalStorageURL_user = model.StringVA('user')
+        self.externalStorageURL_password = model.StringVA('password')
```

Comment:
@Sabrina85 What are the resitrictions on this input? or do we not check that for security reasons? (not showing the range possible options for brute force attacks)

Reply:
Good question. I sent Robert an email asking for this.

Reply:
Robert says this depends in the end on the FTP server used. He proposes to allow numeric and alphanumeric characters only: a..z, A..Z, 0..9